File: ios/renderer/ParagraphRenderer.m
#import "ParagraphRenderer.h"
#import "FontUtils.h"
#import "MarkdownASTNode.h"
#import "ParagraphStyleUtils.h"
#import "RendererFactory.h"
#import "StyleConfig.h"

@implementation ParagraphRenderer

- (id)takeContextSnapshot:(RenderContext *)context
{
  return [context snapshotScope];
}

- (void)renderNodeContent:(MarkdownASTNode *)node
                     into:(NSMutableAttributedString *)output
                  context:(RenderContext *)context
{
  // Synthetic wrapper around tight list item inlines (see wrapListItemInlineRuns
  // in MD4CParser) — render transparently, the list owns spacing and layout.
  if ([node.attributes[@"tight"] boolValue]) {
    [_rendererFactory renderChildrenOfNode:node into:output context:context];
    return;
  }

  // Only set block style if a parent element (e.g. List, Blockquote) hasn't already established one
  BOOL isTopLevel = (context.currentBlockType == BlockTypeNone);

  if (isTopLevel) {
    [context setBlockStyle:BlockTypeParagraph font:_config.paragraphFont color:_config.paragraphColor headingLevel:0];
  }

  NSUInteger start = output.length;
  BOOL shouldApplyMargin =
      (context.currentBlockType == BlockTypeNone || context.currentBlockType == BlockTypeParagraph);

  // Detect if the paragraph is a wrapper for a standalone image to use image-specific spacing
  BOOL isBlockImage = (node.children.count == 1 && ((MarkdownASTNode *)node.children[0]).type == MarkdownNodeTypeImage);
  CGFloat marginTop = isBlockImage ? _config.imageMarginTop : _config.paragraphMarginTop;

  NSUInteger contentStart = start;

  // Handle leading margin for the first element in the document (Index 0 check)
  if (shouldApplyMargin && start == 0) {
    NSUInteger offset = applyBlockSpacingBefore(output, 0, marginTop);
    contentStart += offset;
    start += offset;
  }

  [_rendererFactory renderChildrenOfNode:node into:output context:context];

  if (output.length <= start)
    return;
  NSRange range = NSMakeRange(start, output.length - start);

  // Avoid standard line height on block images to prevent vertical alignment issues
  if (!isBlockImage) {
    applyLineHeight(output, range, _config.paragraphLineHeight);
    applyBaselineOffset(output, range);
  }

  applyTextAlignment(output, range, _config.paragraphTextAlign);

  // Skip marginTop for the first block — already handled by applyBlockSpacingBefore above
  if (shouldApplyMargin && contentStart != 1) {
    NSUInteger inserted = applyParagraphSpacingBefore(output, range, marginTop);
    start += inserted;
  }

  CGFloat marginBottom = 0;
  if (shouldApplyMargin) {
    marginBottom = isBlockImage ? _config.imageMarginBottom : _config.paragraphMarginBottom;
  }

  applyParagraphSpacingAfter(output, start, marginBottom);
}

@end

File: packages/enriched-markdown-ios/Sources/EnrichedMarkdown/Rendering/Renderers/ParagraphRenderer.swift
import UIKit

final class ParagraphRenderer: NodeRenderer {
    private let factory: RendererFactory
    private let config: MarkdownStyleConfig

    init(factory: RendererFactory, config: MarkdownStyleConfig) {
        self.factory = factory
        self.config = config
    }

    func render(node: MarkdownASTNode, into output: NSMutableAttributedString, context: RenderContext) {
        let isTopLevel = context.currentBlockType == .none
        let paragraphStyle = config.paragraph

        if isTopLevel {
            let font = paragraphStyle.font ?? UIFont.preferredFont(forTextStyle: .body)
            let color = paragraphStyle.foregroundColor ?? UIColor.label
            context.setBlockStyle(font: font, color: color)
        }

        let start = output.length
        let shouldApplyMargin = context.currentBlockType == .none || context.currentBlockType == .paragraph
        let isBlockImage = node.children.count == 1 && node.children[0].type == .image
        let imageStyle = config.image
        let marginTop = isBlockImage
            ? (imageStyle.marginTop ?? paragraphStyle.marginTop ?? 0)
            : (paragraphStyle.marginTop ?? 0)
        var contentStart = start

        if shouldApplyMargin, start == 0, marginTop > 0 {
            let offset = ParagraphStyleHelpers.applyBlockSpacingBefore(
                to: output,
                at: 0,
                marginTop: marginTop
            )
            contentStart += offset
        }

        if isBlockImage {
            context.rendersBlockImage = true
        }

        if isTopLevel {
            factory.renderChildren(of: node, into: output, context: context)
            context.clearBlockStyle()
        } else {
            factory.renderChildren(of: node, into: output, context: context)
            if context.currentBlockType == .blockquote {
                ParagraphStyleHelpers.ensureTrailingNewline(in: output)
            }
        }

        context.rendersBlockImage = false

        guard output.length > start else { return }

        let range = NSRange(location: start, length: output.length - start)

        if !isBlockImage, let lineHeight = paragraphStyle.lineHeight {
            // A capped line height would clip taller-than-text plugin content.
            ParagraphStyleHelpers.applyBlockLineHeight(
                to: output,
                range: range,
                lineHeight: lineHeight,
                capMaximum: !ParagraphStyleHelpers.rangeContainsNaturalHeightAttachment(in: output, range: range)
            )
        }

        if let alignment = paragraphStyle.textAlignment {
            ParagraphStyleHelpers.applyTextAlignment(to: output, range: range, alignment: alignment)
        }

        var spacingStart = start
        if shouldApplyMargin, contentStart != 0 {
            spacingStart += ParagraphStyleHelpers.applyParagraphSpacingBefore(
                to: output,
                range: range,
                marginTop: marginTop
            )
        }

        let marginBottom = isBlockImage
            ? (imageStyle.marginBottom ?? paragraphStyle.marginBottom)
            : paragraphStyle.marginBottom

        if shouldApplyMargin, let marginBottom {
            ParagraphStyleHelpers.applyParagraphSpacingAfter(
                to: output,
                at: spacingStart,
                marginBottom: marginBottom
            )
        }
    }
}
